File: February2021/convert-bst-to-greater-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Integer greaterNodeSum = 0;

    // TC : O(n)
    // SC : O(logH)
    public TreeNode convertBST(TreeNode root) {
        if(root==null){
            return null;
        }
        convertBST(root.right);
        root.val = root.val + greaterNodeSum;
        greaterNodeSum = root.val;
        convertBST(root.left);
        return root;

    }
}

File: April2022/C++/Convert_BST_to_Greater_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 Problem Link - https://leetcode.com/problems/convert-bst-to-greater-tree/
 */
class Solution {
public:
    int greaterNodeSum = 0;
    TreeNode* convertBST(TreeNode* root) {
        
        if(root == NULL)
            return NULL;
        convertBST(root -> right);
        greaterNodeSum += root -> val;
        root -> val = greaterNodeSum;
        convertBST(root -> left);
        return root;
    }
};
